assets/js/atomic/blocks/product-elements/add-to-cart/product-types/variable/variation-attributes/attribute-picker.js

```diff
@@ -19,86 +20,72 @@ import {
 	const [ variationId, setVariationId ] = useState( 0 );
-
-	// @todo Support default selected attributes in Variation Picker.
-	const [ selectedAttributes, setSelectedAttributes ] = useState( [] );
+	const [ selectedAttributes, setSelectedAttributes ] = useState( {} );
+	const { dispatchActions } = useAddToCartFormContext();
 
 	const attributeNames = Object.keys( attributes );
-	const hasSelectedAttributes =
-		Object.values( selectedAttributes ).filter( Boolean ).length > 0;
 	const hasSelectedAllAttributes =
 		Object.values( selectedAttributes ).filter(
 			( selected ) => selected !== ''
 		).length === attributeNames.length;
 
-	// Gets valid attribute options for the picker taking current selections into account.
+	// Get options for each attribute picker.
 	const filteredAttributeOptions = useMemo( () => {
-		const options = [];
+		return getActiveSelectControlOptions(
+			attributes,
+			variationAttributes,
+			selectedAttributes
+		);
+	}, [ attributes, variationAttributes, selectedAttributes ] );
 
-		attributeNames.forEach( ( attributeName ) => {
-			const currentAttribute = attributes[ attributeName ];
-			const attributeNamesExcludingCurrentAttribute = attributeNames.filter(
-				( name ) => name !== attributeName
-			);
-			const matchingVariationIds = hasSelectedAttributes
-				? getVariationsMatchingSelectedAttributes( {
-						selectedAttributes,
-						variationAttributes,
-						attributeNames: attributeNamesExcludingCurrentAttribute,
-				  } )
-				: null;
-			const validAttributeTerms =
-				matchingVariationIds !== null
-					? matchingVariationIds.map(
-							( varId ) =>
-								variationAttributes[ varId ][ attributeName ]
-					  )
-					: null;
-			options[ attributeName ] = getSelectControlOptions(
-				currentAttribute.terms,
-				validAttributeTerms
+	// Select variation when selections change.
+	useEffect( () => {
+		if ( hasSelectedAllAttributes ) {
+			setVariationId(
+				getVariationMatchingSelectedAttributes(
+					attributes,
+					variationAttributes,
+					selectedAttributes
+				)
 			);
-		} );
-
-		return options;
+		}
 	}, [
 		attributes,
 		variationAttributes,
-		attributeNames,
 		selectedAttributes,
-		hasSelectedAttributes,
+		hasSelectedAllAttributes,
 	] );
 
-	// Select variation when selections change.
+	// Unset variation when form is incomplete.
 	useEffect( () => {
-		if ( ! hasSelectedAllAttributes ) {
+		if ( ! hasSelectedAllAttributes && variationId > 0 ) {
 			setVariationId( 0 );
-			return;
 		}
+	}, [ hasSelectedAllAttributes, variationId ] );
 
-		const matchingVariationIds = getVariationsMatchingSelectedAttributes( {
-			selectedAttributes,
-			variationAttributes,
-			attributeNames,
+	// Set requests params as variation ID and data changes.
+	useEffect( () => {
+		dispatchActions.setRequestParams( {
+			id: variationId,
+			variation: Object.keys( selectedAttributes ).map(
+				( attributeName ) => {
+					return {
+						attribute: attributeName,
```

Comment:
This is passed to the API. The API doesn't camel case props and only needs key value pairs.

Reply:
Would you prefer we had a setX for each key in the API schema? e.g. setRequestID setRequestVariation and so on.

Reply:
I don't think we need `setX` keys, typically schema's just describe the property and the http method describes what is done with those properties. `setX` feels more like RPC endpoint behaviour.

My comment here is less about the REST API schema shape and more about how the property is stored in state. It sounds like you might be intending on mapping the state 1:1 with the REST API schema, which has value as long as it's consistently applied. Otherwise, I don't see anything wrong with having a layer that takes the data in state and maps that to the shape the REST endpoint requires when prepping for the request.

Reply:
I did indeed intend to map 1:1 with the REST API schema since this schema should be documented and be easiest to work with. I'm also assuming we may allow extensions to add to that schema in the future which is why a 1:1 might make sense—it would be weird to rename some props from state and not others?

I think if the issue is with the _schema_ naming we should tackle in a follow up, although personally I'm happy with the key value pairs of Attribute e.g. _Color_ and Value e.g. _Blue_.